0000-benchmark-suite.md

```diff
@@ -0,0 +1,782 @@
+
+### Randomizing Benchmark Execution Order
+
+If we are running benchmarks with two different versions of Wasmtime, version A
+and version B, we should not take all samples from version A and then all
+samples from version B. We should, instead, intersperse executions of version A
+and version B. While we should always ensure that the CPU frequency scaling
+governor is set to "performance" when running benchmarks, this additionally
+helps us avoid some measurement bias from CPU state transitions that aren't
+constrained within the duration of process execution, like dynamic CPU
+throttling due to overheating.
+
+## Analysis
+
+The worst kind of analysis we can do is something like compare averages with and
+without our proposed optimization, and use that to conclude whether the
+optimization "works". Or, equivalently, eyeball two bars in a graph and decide
+which one looks better. We don't know whether the difference is meaningful or
+just noise.
+
+A better approach is to use a test of statistical significance to answer the
+question as to whether the difference is meaningful or not. For comparing two
+means we would use [Student's
+t-test](https://en.wikipedia.org/wiki/Student%27s_t-test) and for comparing
+three or more means we would use [analysis of variance
+(ANOVA)](https://en.wikipedia.org/wiki/Analysis_of_variance). These tests allow
+us to justify statements like "we are 95% confident that change X produces a
+meaningful difference in execution time".
+
+We aren't, however, solely interested in whether execution time is different or
+not. We want to know how much of a speed up an optimization yields! A
+significance test, while an improvement over eyeballing results, is still
+"asking the wrong question". This leads us to the best approach for analyzing
+performance results: [effect size confidence
+intervals](https://en.wikipedia.org/wiki/Effect_size). This analysis justifies
+statements like "we are 95% confident that change X produces a 5.5% (+/- 0.8%)
+speed up compared to the original system". See the "Challenge of Summarizing
+Results" and "Measuring Speedup" sections of *Rigorous Benchmarking in
+Reasonable Time*<sup>[4](#ref-4)</sup> for further details.
+
+Our benchmark results analysis code should report effect size along with its
+confidence interval. The primary data visualization we produce should be a bar
+graph showing speed up or slowdown normalized to the control, with the
+confidence intervals overlaid. The R language has an [off-the-shelf
+package](https://github.com/easystats/effectsize) for working with effect size
+confidence intervals, but we can also always implement this analysis ourselves
+should we choose another statistics and plotting language or framework.
+
+## Developer Ergonomics
+
+For optimal developer ergonomics, it should be possible to run the benchmark
+suite remotely by leaving a special comment on a github pull request:
+
+> @bytecodealliance-bot bench
```

Comment:
Who should be allowed to run it? Will there be a rate-limit per developer?

Reply:
This is discussed a bit in #3 as well. I expect that we will use an explicit allow list. I don't think we need to worry about rate limiting right off the bat, since the allow list will be people we all trust and can just ping if there are any issues.